src/host/directio.cpp

```diff
@@ -237,35 +138,92 @@ void EventsToUnicode(_Inout_ std::deque<std::unique_ptr<IInputEvent>>& inEvents,
                                                           const std::span<const INPUT_RECORD> buffer,
                                                           size_t& written,
                                                           const bool append) noexcept
+try
 {
     written = 0;
 
+    if (buffer.empty())
+    {
+        return S_OK;
+    }
+
     LockConsole();
     auto Unlock = wil::scope_exit([&] { UnlockConsole(); });
 
-    try
+    const auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+    til::small_vector<INPUT_RECORD, 16> events;
+
+    // Check out the loop below. When a previous call ended on a leading DBCS we store it for
+    // the next call to WriteConsoleInputAImpl to join it with the now available trailing DBCS.
+    if (context.IsWritePartialByteSequenceAvailable())
     {
-        auto events = IInputEvent::Create(buffer);
+        auto lead = context.FetchWritePartialByteSequence(false)->ToInputRecord();
+        const auto& trail = buffer.front();
+
+        if (trail.EventType == KEY_EVENT)
+        {
+            const char narrow[2]{
+                lead.Event.KeyEvent.uChar.AsciiChar,
+                trail.Event.KeyEvent.uChar.AsciiChar,

```

Comment:
hey wait a minute, won't this only work for things that take up 2 code units? a UTF-8 string can be up to 4 :wink:

Reply:
Yep, it works just as bad as it did before! Well... Sort of. At least it now checks if the trailing event is an actual `KEY_EVENT` instead of blindly casting it to a `KeyEvent` struct. 😅

I'm hoping to move this entire block of code into `InputBuffer` at a later time and come up with a proper caching system.